### DISCUSSION TITLE
[Idea] : Calculation of taxes

### DISCUSSION THREAD

**Discussion by RosaBenouamer** [PrestaShop/PrestaShop#34327] [Ideas]
### The problem

1 . We have a group customer named 'Green'
2. we add specific price to all product
3. add order from backoffice to check

Capture d’écran 2023-10-06 à 4 10 00 PM

A - The round mode is line.
See the invoice

![image](https://github.com/PrestaShop/PrestaShop/assets/124665010/1ec85e6b-0f82-499f-8d6a-9f9ade72acdc)

Total Tax not same when we addition tax value (product+shipping) and the result in database

It seems that VAT 5.61 is rounding so 5.61 * 7 = 39.27 because of round mode line
So in Total VAT you are expexted 39.27 + 10.50 = 49.77
but you get a different result because today Total VAT seems to be calculated like rounding of total
So (62.360.0097) = 39.29
So 39,29 + 10.50 = 49.79
So sum of rounding for VAT is different of rounding of sum.

B- the round mode is total
it's ok
![image](https://github.com/PrestaShop/PrestaShop/assets/124665010/d0dac793-cb0d-4ca0-a75c-9cf5c58699ba)

To be coherent we have to use the same rounding mode in taxese details and in Total VAT.

### The solution

I dont' know in the law 
if the TVA has to be calculated exactly not from rounding so we have to changeTotal Tax 39.27 by 39.29 to be coherent or 
if as we are in round line we have to use 39.27 + 10.50 and so have Total Tax 49.77.

In conclusion, should VAT be rounded according to the item, line or total mode group by rate, or should the sum total be rounded?

Are you aware about this point ?

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [ ] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by cruftex** [comment]
Tax from the total is the correct way.

In general, minimise the rounding errors, so only round where needed. Problems you would get with per line rule:

- Consider the different results you would get if you order 100 pieces of the same product in 100 lines, one line or 10 lines with 10 items each. So per line does not give consistent tax calculations. Alternatively, per each item would give consistent results, but this has other problems which are covered in the next bullet point
- Let's consider you have 100 different products that cost 0.01 cents each, would you be allowed to put 0 Euros VAT on the invoice?

**Comment by Patrickclicweb** [comment]
En matière de tva (En france) on est assujetti à partir du total ht donc plusieurs lignes de produit egale un moreprisesT qui ensuite est soumis à la TVA 
et non une ligne de produit plus TVA puis une autre ligne plus tva et ensuite le total des 2 lignes de produit reprise en Ht avec ajout de la tva n'est, il me semble pas conforme !

**Comment by RosaBenouamer** [comment]
Ok in our example we have a unique tax rate.

Imagine we have
Product A1 and A2 with tax rate A quantity A
and 
Product B tax rate B quantity B

Do we agree that we have :
on the one hand to calculate rounded( of sum of TVA for Product A1 and A2) which have the same rate A = RA
on the other hand to calculate rounded (of sum of TVA for Product B) with the other rate B = RB
and after to sum RA and RB ?

Indeed in this part of the Invoice we have Tax group by rate
![image](https://github.com/PrestaShop/PrestaShop/assets/124665010/f1e948b3-8ebd-4896-ab66-f9bcfcf978ed)

**Comment by cruftex** [comment]
@RosaBenouamer 
This is getting confusing and hard for people to follow. It is important to have as simple and minimal examples as possible and concise terminology.

Problems:
- "1 . We have a group customer named 'Green' 2. we add specific price to all product" how is this relevant for the question?
- I think the example of the invoice changed. It had tax incl. unit prices before now it has tax excl. prices
- In the initial issue text you use the terms VAT and TVA. What is TVA?

About the question:
- If TVA is the included VAT in the product price then: no. You need to deduct the VAT invoice value from the invoiced total per tax rate, because: See my first comment

I was writing this in my text  https://github.com/PrestaShop/PrestaShop/discussions/32969 under tax calculation. I just improved the pseudo code a bit:

Depending whether the prices are tax inclusive (B2C) or tax exclusive (B2B), either the tax including position totals
or tax excluding position totals are summed up to the final total.

More formal in pseudo code:

````
for each distinctive VAT-rate:
  unit_price = inclusiveVat ? displayedVatInclusiveUnitPrice : displayedVatExclusiveUnitPrice
  position_total = qty * unit_price
  product_total = position_price1 + position_price2 + ...
  if (inclusveVat) {
     vat_inclusive_sum = product_total
     included_vat = round(product_total / (1 + vat_rate) * vat_rate)
    vat_exclusive_sum = vat_inclusive_sum - included_vat
  } else {
    vat_exclusive_sum = product_total
     included_vat =  round(product_total * vat_rate)
     vat_inclusive_price = vat_exclusive + included_vat
  } 
````

**Comment by RosaBenouamer** [comment]
Hello @cruftex 

> @RosaBenouamer This is getting confusing and hard for people to follow. It is important to have as simple and minimal examples as possible and concise terminology.
> 
> Problems:
> 
> * "1 . We have a group customer named 'Green' 2. we add specific price to all product" how is this relevant for the question?

You're right indeed I just take an example we are talking about with someone else here https://github.com/PrestaShop/PrestaShop/issues/33912#issuecomment-1748782474. I have prefer to create a discussion about it. 
But for the example the group and the specific price are not required for a good understanding.

> * I think the example of the invoice changed. It had tax incl. unit prices before now it has tax excl. prices

Sorry I don't understand when is "before" and when is "now"

> * In the initial issue text you use the terms VAT and TVA. What is TVA?

VAT = TVA sorry (TVA is in French)

> 
> About the question:
> 
> * If TVA is the included VAT in the product price then: no. You need to deduct the VAT invoice value from the invoiced total per tax rate, because: See my first comment
> I was writing this in my text #32969 under tax calculation. I just improved the pseudo code a bit:
> 
> Depending whether the prices are tax inclusive (B2C) or tax exclusive (B2B), either the tax including position totals or tax excluding position totals are summed up to the final total.
> 
> More formal in pseudo code:
> 
> ```
> for each distinctive VAT-rate:
>   unit_price = inclusiveVat ? displayedVatInclusiveUnitPrice : displayedVatExclusiveUnitPrice
>   position_total = qty * unit_price
>   product_total = position_price1 + position_price2 + ...
>   if (inclusveVat) {
>      vat_inclusive_sum = product_total
>      included_vat = round(product_total / (1 + vat_rate) * vat_rate)
>     vat_exclusive_sum = vat_inclusive_sum - included_vat
>   } else {
>     vat_exclusive_sum = product_total
>      included_vat =  round(product_total * vat_rate)
>      vat_inclusive_price = vat_exclusive + included_vat
>   } 
> ```

Do you agree with this invoice or not. If not please correct with the expected results :-)
<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>

Reference |   | Product |   |   | Tax Rate | Base price (Tax exl.) | Unit Price (Tax excl.) |   | Qty |   | Total (Tax excl.)
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
  | Article 1 - Black Balloon |   |   |   | 20 % | 6 | €6 |   | 12 |   | €72
  | Article 2 - White Balloon |   |   |   | 20 % | 100 | €100.1 |   | 2 |   | €200.2
  | Article 3 - Grey Balloon |   |   |   | 21 % | 7 | €7 |   | 2 |   | €14
  | Article 4 - Skipping rope |   |   |   | 21 % | 25 | €25 |   | 10 |   | €250
| -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
|Tax Detail | Tax Rate | Base price | Total Tax |   |   |   | Total Products |   |   |   | €536.2
|Shipping |   |   | €0 |   |   |   | Total Discounts |   |   |   | €0
|Products | 20 % | €272.2 | 54.4 |   |   |   | Shipping Costs |   |   |   |  
  | 21 % | €264 | 55.4 |   |   |   | Total (Tax excl.) |   |   |   | €536.2
  |   |   |   |   |   |   | Total Tax |   |   |   | €109.8
  |   |   |   |   |   |   | Total |   |   |   | €646


</body>
</html>

**Comment by cruftex** [comment]
@RosaBenouamer 
> Do you agree with this invoice or not.

I think we should work with easier examples. Also, ideally the example would trigger incorrect calculations now. What about this:

Invoice liens:

Name | VAT | Price (Tax excl.) | Quantity | Total (Tax excl.)
-- | -- | -- | -- | --
Product 1 | 10% | 0,01 | 99 | 0,99
Product 1 | 10% | 0,01 | 1 | 0,01
Product 2 | 20% | 1,00 | 1 | 1,00

Tax breakdown:

  | VAT | Total | Tax
-- | -- | -- | --
Products | 10% | 1,00 | 0,10
Products | 20% | 1,00 | 0,20

Invoice totals:

Name | Value
-- | -- 
Total Products | 2,00
Shipping | 0,00
Total (Tax excl.) | 2,00
Invoice Total | 2,30

**Reply by RosaBenouamer** [reply]
@cruftex  It totally agree for this case B to B

**Comment by cruftex** [comment]
For tax incl. invoice an easy example would be:

Name | VAT | Price (Tax incl.) | Quantity | Total (Tax incl)
-- | -- | -- | -- | --
Product 1 | 10 % | 0,01 | 109 | 1,09
Product 1 | 10 % | 0,01 | 1 | 0,01
Product 2 | 20 % | 1,20 | 1 | 1,20

Tax breakdown:

  | VAT | Total (Tax incl.) | Tax
-- | -- | -- | --
Products | 10 % | 1,10 | 0,10
Products | 20 % | 1,20 | 0,20

Invoice totals:

Name | Value
-- | --
Total Products | 2,30
Shipping | 0,00
Total (Tax excl.) | 2,00
Invoice Total | 2,30

**Reply by RosaBenouamer** [reply]
@cruftex I understand your mind but it's linked with the following discussion : https://github.com/PrestaShop/PrestaShop/discussions/34262.
Your solution (corresponding to solution 3 in the discussion) is based on the price with VAT displayed for the product which could make sens to be coherent in B to C but IMO is not coherent if a client appear in two groups one B to B and the other one B to C and try to buy the same product. Thats' why the solution 2 in the discussion seems so be interesting. I have send the question to an accountable tu decide among the 3 solutions. I a waiting for his feed back. Hope soon :-)

**Comment by RosaBenouamer** [comment]
@cruftex : Why for VAT 10% you get 0.10 instead of 0.11 and 0.20 instead of 0.24 ? Not sure to understand

**Comment by cruftex** [comment]
@RosaBenouamer 

> Why for VAT 10% you get 0.10 instead of 0.11 and 0.20 instead of 0.24 ? Not sure to understand

Maybe good we have a simple example. 1.00 Euro (VAT excl.) plus 10% VAT (1.00 * 10% = 0.10) is 1.10 Euro
